### ISSUE TITLE
Wrong platform tag with `python -m build .`

### ISSUE THREAD

**Issue by jeandet** [FFY00/meson-python#124]
Hello,

While I wanted to investigate more the tag issue I was talking on #118, I discover that calling [build_wheel](https://github.com/FFY00/meson-python/blob/main/mesonpy/__init__.py#L934) manually produces the right wheel name (`pycdfpp-0.3.2-cp310-cp310-linux_x86_64.whl` for example) while `python -m build .` produces `pycdfpp-0.3.2-cp310-cp310-any.whl`.
I don't know enough about pep517 and build module to investigate further...
Any idea what could be the cause here or what I could do to help?

**Comment by rgommers** [comment]
@jeandet thanks for the report. Can you please add more details? Is it a matter of cloning https://github.com/SciQLop/CDFpp on 64-bit Linux and running `python -m build .` to reproduce the problem?

Your `pyproject.toml` is missing a few dependencies it looks like, it only lists `numpy` but the build definition has:
```
pybind11_dep = dependency('pybind11', required : true, fallback:['pybind11','pybind11_dep'])
hedley_dep = dependency('hedley', main : true, fallback : ['hedley', 'hedley_dep'])
```

**Comment by jeandet** [comment]
@rgommers, yes it only requires a C++ compiler, pybind11 and hedley are provided by meson's wrap database.
I just did a test from a fresh ubuntu container:
```
podman run -it --rm ubuntu:20.04 bash

apt update
apt install g++ python3-pip git meson python3.8-venv patchelf
pip3 install build
git clone  https://github.com/SciQLop/CDFpp 
cd CDFpp/
python3 -m build .
```
And got `Successfully built pycdfpp-0.3.2.tar.gz and pycdfpp-0.3.2-py3-none-any.whl`

I also suspect that any C or C++ extension module would have the same issue.

**Comment by jeandet** [comment]
This seems to be close to #104 and #95 maybe my bug is a duplicate of #104?

**Comment by jeandet** [comment]
Another weird update, on GH Actions it doesn't work :/ https://github.com/SciQLop/CDFpp/actions/runs/2859854274.
I did more tests, it seems that, calling `build_wheel` directly only generates a correct name `pycdfpp-0.3.2-cp310-cp310-linux_x86_64.whl` on Fedora (native or from a container) while I tried on an Ubuntu container and it produces `pycdfpp-0.3.2-py3-none-any.whl`.

**Comment by rgommers** [comment]
There's a separate code path for Debian and derivatives, because Debian messes up `sysconfig` & co too much: https://github.com/FFY00/meson-python/blob/fc29c1bc820f4b4abbb929aa68d90cf6e6b40a0c/mesonpy/__init__.py#L240-L249

**Comment by jeandet** [comment]
Ok, I see, some part of the code looks quite convoluted to me (a least a first look). I wonder why it doesn't uses a much simpler and robust (I guess) solution looking for binaries either in `intro-install_plan.json` or for `shared modules` with `"installed" :true" in `intro-targets.json`. It seems safe to assume that as long as there is a single binary installed then the lib is impure (ignoring some very marginal corner cases like pure python lib shipping with some firmware).

**Comment by FFY00** [comment]
Hi, thank you for opening this, and sorry for the delay getting back to you.

The reason we don't simply check if the file is a binary is because you can actually ship binaries in purelib as data. One good example for this are the entrypoint script runners on Windows (see https://github.com/pypa/installer/tree/main/src/installer/_scripts).

The current code is a bit complex indeed, I have been meaning to refactor it. The root of your issue is that the code that checks if the file should be taken in consideration when calculating the ABI tag gets messed up due to Debian's patching of Python. There were two codepaths, in different classes, doing this essentially, which is not optimal. In #130 I refactored it back to one path, so now the ABI detection should be more reliable, as it re-uses the existing mapping code.

**Comment by jeandet** [comment]
@FFY00 I've made this little [repo](https://github.com/jeandet/meson_cpython) to reproduce the issue with a minimal set of files, you can copy it in meson-python tests if you want.

**Comment by dnicolodi** [comment]
I'm encountering the same problem on Ubuntu and macOS. This is the project https://github.com/dnicolodi/python-siphash24 If you want to have a look at the details of the compilation it should be enough to fork the project and push a dummy commit to trigger Github Actions. I'm happy to provide more details and to help fix this, but at the moment I don't know where to start.

**Comment by dnicolodi** [comment]
Same problem on Windows with MinGW-W64 too. MSVC fails at an earlier stage for what I think is a bug in Meson https://github.com/mesonbuild/meson/issues/10775 I would have expected things to be a bit smoother given that SciPy is using Meson and meson-python for building its wheels.

**Comment by dnicolodi** [comment]
In my case the issue stems from `mesonpy.Project.is_pure()` returning `True` when it should return `False`. This is because this method checks whether any of `{bindir}`, `{libdir_shared}`, `{libdir_static}`, `{py_platlib}`, `{moduledir_shared}` is in one of the installation destination for the project build artifacts as exposed by `$build/meson-info/build/meson-info/intro-install_plan.json`, however, in my case the install paths are fully resolved directories: `{'targets': {'/Users/daniele/src/python-siphash24/.mesonpy-lxc7bf0m/build/siphash24.cpython-310-darwin.so': {'destination': '/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/siphash24.cpython-310-darwin.so', 'tag': 'runtime'}}}`.

**Comment by dnicolodi** [comment]
The full path in `$build/meson-info/build/meson-info/intro-install_plan.json` seems a bug in Meson. If I add a dummy shared library, it appears with the correct `{libdir_shared}` destination path prefix.

**Comment by rgommers** [comment]
> @FFY00 I've made this little [repo](https://github.com/jeandet/meson_cpython) to reproduce the issue with a minimal set of files, you can copy it in meson-python tests if you want.

Thanks @jeandet. I can reproduce the issue on Arch Linux. Cloning your repo and running `python -m build .` yields a wheel with a `any` tag and metadata:
```
Root-Is-Purelib: true
Tag: cp310-cp310-any
```
while the binary goes into `platlib`:

```
$ tree dist
dist
├── meson_cpython-1.0.0-cp310-cp310-any.whl
├── meson_cpython-1.0.0.data
│   └── platlib
│       └── meson_cpython.cpython-310-x86_64-linux-gnu.so
├── meson_cpython-1.0.0.dist-info
│   ├── METADATA
│   ├── RECORD
│   └── WHEEL
└── meson_cpython-1.0.0.tar.gz

3 directories, 6 files
```

That reproduces with:
```
# Name                    Version                   Build  Channel
meson                     0.63.2                   pypi_0    pypi
meson-python              0.8.1                    pypi_0    pypi
```
_Important (the usual footgun for testing): if one doesn't disable build isolation, the versions in the active environment aren't the ones used for the build._

I also checked that the root cause here is an absolute path instead of `{py_platlib}` in `intro-install_plan.json`.
```
{"targets": {"/home/rgommers/code/tmp/testcase/meson_cpython/build/meson_cpython.cpython-310-x86_64-linux-gnu.so": {"destination": "/home/rgommers/anaconda3/envs/scipy-dev/lib/python3.10/site-packages/meson_cpython.cpython-310-x86_64-linux-gnu.so", "tag": "runtime"}}}
```

I then upgraded `meson` to latest `master`, and built with `python -m build . --no-isolation -Cbuilddir=build`. Then the problem is fixed:
```
$ ls dist/
meson_cpython-1.0.0-cp310-cp310-linux_x86_64.whl
meson_cpython-1.0.0.tar.gz
```
And `intro-install_plan.json` contains:
```
{"destination": "{py_platlib}/meson_cpython.cpython-310-x86_64-linux-gnu.so"
```

So tl;dr this is fixed in Meson `master`, and _perhaps_ will be backported to `0.63.3` - but is broken in `0.63.2`.

I can also reproduce this with the existing test in `tests/packages/purelib-and-platlib/`, which produces `-any.whl` with meson `0.63.2` and `-cp310-cp310-linux_x86_64.whl` with meson `master`.

tl;dr we are good here once the fix is in a released Meson version; we should then also set the minimum required Meson version to that release.

**Comment by dnicolodi** [comment]
For who is following along, the Meson issue is tracked here https://github.com/mesonbuild/meson/issues/10601 It is indeed resolved in meson master. My problems reported above are solved by using meson master.
